### PR TITLE
Add the Streaming Random Patches ensemble classifier

### PR DESCRIPTION
**Pull request by jacobmontiel**
This PR covers:

- Adding the Streaming Random Patches classifier
- Adding support in the confusion matrix for corner cases when mixed data types are passed.

Test notes: We skip the `check_shuffle_features_no_impact` test as it is expected that the order of the features impacts this method.

### CODE REVIEW

File: river/ensemble/streaming_random_patches.py [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+
+    Parameters
+    ----------
+    model
+        The base estimator.
+    n_models
+        Number of members in the ensemble.
+    subspace_size
+        Number of features per subset for each classifier where `M` is the
+        total number of features.<br/>
```
Comment (smastelini) [inline river/ensemble/streaming_random_patches.py:33]: I think I need start using these line breaks in my PRs XD

File: river/ensemble/streaming_random_patches.py [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+            created_on=n_samples_seen,
+            disable_background_learner=self.disable_background_learner,
+            base_drift_detector=self.base_drift_detector,
+            base_warning_detector=self.base_warning_detector,
+            is_background_learner=True,
+            features=subspace,
+            nominal_attributes=self.nominal_attributes,
+            rng=self._rng
+        )
+        # TODO Hard-reset the warning method
```
Comment (smastelini) [inline river/ensemble/streaming_random_patches.py:435]: Is this still valid?

**Review by smastelini — CHANGES_REQUESTED** [review]
Great work @jacobmontiel! SRPClassifier is a must in river! I finished a first, and it's looking pretty good already. I left some comments that may deserve discussion. Do not hesitate to contact me if you have any questions.

File: river/ensemble/streaming_random_patches.py [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+        The training method to use.<br/>
+        * 'subspaces' - Random subspaces.<br/>
+        * 'resampling' - Resampling.<br/>
+        * 'patches' - Random patches.
+    lam
+        Lambda value for resampling.
+    drift_detector
+        Drift detector. If `None`, disables concept drift detection and the background learner.
+    warning_detector
+        Warning detector.
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py:50]: No, they are different. They have different sensitivity to concept drift. The warning detector is slightly more sensitive to changes, thus "warning". The drift detector is less sensitive in order to avoid false detections.

File: river/ensemble/streaming_random_patches.py [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+        self.metric = metric
+        self.nominal_attributes = nominal_attributes if nominal_attributes else []
+        self.seed = seed
+        # self._random_state is the actual object used internally
+        self._rng = check_random_state(self.seed)
+
+        self._n_samples_seen = 0
+        self._subspaces = None
+
+        self._base_learner_class = StreamingRandomPatchesBaseLearner
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py:136]: Because this way we can change the base class for regression ;-)

File: river/ensemble/streaming_random_patches.py [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+
+    @property
+    def _wrapped_model(self):
+        return self.model
+
+    def learn_one(self, x: dict, y: base.typing.ClfTarget, **kwargs):
+        self._n_samples_seen += 1
+
+        if not self.models:
+            self._init_ensemble(list(x.keys()))
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py:146]: We will have to trust the user on this one. This method, like many others, does not support feature evolution. What you see on the first call is what is assumed to be the feature space for initializing the ensemble (this function is only called once).
However, there is a middle ground option. When there is drift, a new "random" subspace is selected. We could use a set of "seen features" in this case to account for everything we have seen so far.

File: river/ensemble/streaming_random_patches.py [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+        Lambda value for resampling.
+    drift_detector
+        Drift detector. If `None`, disables concept drift detection and the background learner.
+    warning_detector
+        Warning detector.
+    disable_weighted_vote
+        If True, disables weighted voting.
+    disable_background_learner
+        If True, disables background learner and trees are reset
+        if drift is detected.
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py]: This was bothering me as well. There is indeed a way to achieve this. Fixed in d0dc434f8247752e226d357231df628681ad33cf

Now the background learner is deactivated if `warning_detector is None`

File: river/ensemble/streaming_random_patches.py [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+    nominal_attributes
+        List of Nominal attributes. If empty, then assumes that all
+        attributes are numerical.
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py]: Added a note in d0dc434f8247752e226d357231df628681ad33cf

File: river/ensemble/streaming_random_patches.py [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+
+    _FEATURES_SQRT = "sqrt"
+    _FEATURES_SQRT_INV = "rmsqrt"
+
+    _VALID_TRAINING_METHODS = {_TRAIN_RANDOM_PATCHES, _TRAIN_RESAMPLING, _TRAIN_RESAMPLING}
+
+    def __init__(self, model=HoeffdingTreeClassifier(grace_period=50,
+                                                     split_confidence=0.01),
+                 n_models: int = 100,
+                 subspace_size: int = .6,
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py]: Updated the type in d0dc434f8247752e226d357231df628681ad33cf

File: river/ensemble/streaming_random_patches.py [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+        self.training_method = training_method
+        self.lam = lam
+        self.drift_detector = drift_detector
+        self.warning_detector = warning_detector
+        self.disable_weighted_vote = disable_weighted_vote
+        self.disable_background_learner = disable_background_learner
+        self.metric = metric
+        self.nominal_attributes = nominal_attributes if nominal_attributes else []
+        self.seed = seed
+        # self._random_state is the actual object used internally
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py]: Removed in d0dc434f8247752e226d357231df628681ad33cf

File: river/ensemble/streaming_random_patches.py [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+        if not self.models:
+            self._init_ensemble(list(x.keys()))
+
+        for model in self.models:
+            # Get prediction for instance
+            y_pred = model.predict_proba_one(x)
+            if y_pred:
+                y_pred = max(y_pred, key=y_pred.get)
+            else:
+                # Model is empty, default to class 0
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py]: Updated in d0dc434f8247752e226d357231df628681ad33cf

File: river/ensemble/streaming_random_patches.py [hunk 11]
```diff
@@ -0,0 +1,478 @@
...
+
+        if not self.models:
+            self._init_ensemble(features=list(x.keys()))
+            return y_pred
+
+        for model in self.models:
+            y_proba_temp = model.predict_proba_one(x)
+            metric_value = model.metric.get()
+            if not self.disable_weighted_vote and metric_value > 0.:
+                y_proba_temp = {k: val * metric_value for k, val in y_proba_temp.items()}
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py:186]: This is a valid concern. I think for classification we are safe with the current metrics. However, the point is interesting although outside of the scope of this PR.

File: river/ensemble/streaming_random_patches.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py:144]: On second thought, the issue is still going to be there. If an incoming sample is missing any of the "seen features" there will be an error when trying to find it. We could protect the code for this corner case, however, it could impact performance as the model would be learning on fewer features than intended.

File: river/ensemble/streaming_random_patches.py [hunk 12]
```diff
@@ -0,0 +1,478 @@
...
+                    k = int(np.round(n_features * percent)) + 1
+            elif isinstance(self.subspace_size, int) and self.subspace_size > 2:
+                # k is a fixed number of features
+                k = self.subspace_size
+            elif self.subspace_size == self._FEATURES_SQRT:
+                # k is sqrt(M)+1
+                k = int(np.round(np.sqrt(n_features)) + 1)
+            elif self.subspace_size == self._FEATURES_SQRT_INV:
+                # k is M-(sqrt(M)+1)
+                k = n_features - int(np.round(np.sqrt(n_features)) + 1)
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py]: Nice catch! Changed in d0dc434f8247752e226d357231df628681ad33cf

File: river/ensemble/streaming_random_patches.py [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+            else:
+                # k == 0 or k > n_features (subspace size is larger than the
+                # number of features), then default to re-sampling
+                self.training_method = self._TRAIN_RESAMPLING
+
+    def _init_ensemble(self, features: list):
+        self._generate_subspaces(features=features)
+
+        # # Reset the base estimator for safety.
+        # self.model.reset()
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py]: Actually, I forgot to refactor this. It is fixed in d0dc434f8247752e226d357231df628681ad33cf

File: river/ensemble/streaming_random_patches.py [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+        self.features = features
+
+        # Drift and warning detection
+        self.base_drift_detector = base_drift_detector  # Drift detector prototype
+        self.base_warning_detector = base_warning_detector  # Warning detector prototype
+
+        self.disable_background_learner = disable_background_learner
+
+        if base_drift_detector is not None:
+            self.disable_drift_detector = True
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py]: My bad, this is indeed incorrect. Fixed in d0dc434f8247752e226d357231df628681ad33cf

File: river/ensemble/streaming_random_patches.py [hunk 15]
```diff
@@ -0,0 +1,478 @@
...
+                    f"Valid options are: int [2, M], float (0., 1.],"
+                    f" {self._FEATURES_SQRT}, {self._FEATURES_SQRT_INV}"
+                )
+            if k < 0:
+                # k is negative, calculate M - k
+                k = n_features + k
+
+            # Generate subspaces. The subspaces is a 2D matrix of shape
+            # (n_estimators, k) where each row contains the k feature indices
+            # to be used by each estimator.
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py]: The comment is still valid. I just did some small corrections in d0dc434f8247752e226d357231df628681ad33cf

File: river/ensemble/streaming_random_patches.py (hunk 2 above)
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py]: Actually, I forgot to refactor tyhis. It is fixed in d0dc434f8247752e226d357231df628681ad33cf

File: river/ensemble/streaming_random_patches.py [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+                metric=copy.deepcopy(self.metric),
+                created_on=self._n_samples_seen,
+                base_drift_detector=self.drift_detector,
+                base_warning_detector=self.warning_detector,
+                is_background_learner=False,
+                features=subspace,
+                nominal_attributes=self.nominal_attributes,
+                rng=self._rng))
+
+    def reset(self):
```
Comment (MaxHalford) [inline river/ensemble/streaming_random_patches.py:286]: Is there need for a bespoke `reset` method, or can we handle this generically with a function in `utils`?

File: river/ensemble/streaming_random_patches.py [hunk 17]
```diff
@@ -0,0 +1,478 @@
...
+            # Note: Pass the original instance x so features are correctly
+            # selected based on the corresponding subspace
+            self._background_learner.learn_one(x=x, y=y, sample_weight=sample_weight,
+                                               n_samples_seen=n_samples_seen, rng=rng)
+
+        if not self.disable_drift_detector and not self.is_background_learner:
+            correctly_classifies = self.model.predict_one(x_subset) == y
+            # Check for warnings only if the background learner is active
+            if not self.disable_background_learner:
+                # Update the warning detection method
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py:387]: I see the point. Perhaps the note should be in the `metric` parameter description?

File: river/ensemble/streaming_random_patches.py [hunk 18]
```diff
@@ -0,0 +1,478 @@
...
+
+        total = sum(y_pred.values())
+        if total > 0:
+            return {label: proba / total for label, proba in y_pred.items()}
+        return y_pred
+
+    def _generate_subspaces(self, features: list):
+        n_features = len(features)
+
+        self._subspaces = [None] * n_features
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py:197]: This way is faster ;-)

```
%timeit [None] * 100
331 ns ± 5.27 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
%timeit [None for _ in range(100)]
2.83 µs ± 75.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

File: river/ensemble/streaming_random_patches.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/ensemble/streaming_random_patches.py]: Understood!

File: river/ensemble/streaming_random_patches.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/ensemble/streaming_random_patches.py]: Cool!

File: river/ensemble/streaming_random_patches.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/ensemble/streaming_random_patches.py:134]: Nice trick! Hahaha

I am convinced :D

File: river/ensemble/streaming_random_patches.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/ensemble/streaming_random_patches.py:186]: You are right. That a problem for another day. We will revisit that later with SRPRegressor.

File: river/ensemble/streaming_random_patches.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/ensemble/streaming_random_patches.py:144]: That is a valid concern. I think emerging and missing features should be more explored within river. We could add a set of tests to evaluate such scenarios for all the available estimators.

I vote to protect the code against this potential error. In the case of SRPClassifier it should be simpler than, let say, doing the same thing with the current implementations of k-NN.

File: river/ensemble/streaming_random_patches.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/ensemble/streaming_random_patches.py:197]: Perfect! My failed argument was related to speed. I am more than convinced XD

This will change my coding style in some situations.

File: river/ensemble/streaming_random_patches.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/ensemble/streaming_random_patches.py]: Cool!

File: river/ensemble/streaming_random_patches.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/ensemble/streaming_random_patches.py:387]: That's a valid point. Is the metric only use to weight the predictions?

**Review by smastelini — APPROVED** [review]
This PR is ready to be merged. There is a concern related to missing `(key, value)` pairs in the incoming instances. But since this problem might affect multiple learners, we are going to address the issue in other PRs.

@jacobmontiel and @MaxHalford, we need to create a suite of tests to check the robustness of learners against missing key, value pairs in `x`.

File: river/ensemble/streaming_random_patches.py (hunk 16 above)
Comment (jacobmontiel) [reply river/ensemble/streaming_random_patches.py:286]: I left it for reference. If we go with a generic function then we can remove it.
